packages/node-core/src/indexer/connectionPool.service.ts

```diff
@@ -1,66 +1,167 @@
+
+const MAX_FAILURES = 5;
+const LOG_INTERVAL_MS = 60 * 1000; // Log every 60 seconds
+const RESPONSE_TIME_WEIGHT = 0.7;
+const FAILURE_WEIGHT = 0.3;
+const RETRY_DELAY = 60 * 1000;
 
 export interface ApiConnection {
-  apiConnect(): Promise<void>;
-  apiDisconnect(): Promise<void>;
+  //apiConnect(): Promise<void>;
+  //apiDisconnect(): Promise<void>;
+  api: any;
+}
+
+export enum ApiErrorType {
+  Timeout = 'timeout',
+  Connection = 'connection',
+  Default = 'default',
+}
+
+export class ApiConnectionError extends Error {
+  errorType: ApiErrorType;
+
+  constructor(name: string, message: string, errorType: ApiErrorType) {
+    super(message);
+    this.name = name;
+    this.errorType = errorType;
+  }
+}
+
+interface ConnectionPoolItem<T> {
+  endpoint: string;
+  connection: T;
+  performanceScore: number;
+  backoffDelay: number;
+  failureCount: number;
 }
 
 @Injectable()
-export class ConnectionPoolService<T extends ApiConnection> implements OnApplicationShutdown {
+export class ConnectionPoolService<T extends IApi<any, any, any> & IApiConnectionSpecific<any, any, any>>
+  implements OnApplicationShutdown
+{
   private allApi: T[] = [];
-  private connectionPool: Record<number, T> = {};
-  private taskCounter = 0;
+  private apiToIndexMap: Map<T, number> = new Map();
+  private pool: Record<number, ConnectionPoolItem<T>> = {};
+
+  private errorTypeToScoreAdjustment = {
+    [ApiErrorType.Timeout]: -10,
+    [ApiErrorType.Connection]: -20,
+    [ApiErrorType.Default]: -5,
+  };
 
   async onApplicationShutdown(): Promise<void> {
-    await Promise.all(
-      Object.keys(this.connectionPool)?.map((key) => this.connectionPool[toNumber(key)].apiDisconnect())
-    );
+    await Promise.all(Object.values(this.pool).map((poolItem) => poolItem.connection.apiDisconnect()));
   }
 
-  addToConnections(api: T): void {
+  addToConnections(api: T, endpoint: string): void {
+    const index = this.allApi.length;
     this.allApi.push(api);
-    this.connectionPool[this.allApi.length - 1] = api;
+
+    const poolItem: ConnectionPoolItem<T> = {
+      connection: api,
+      performanceScore: 100,
+      failureCount: 0,
+      endpoint: endpoint,
+      backoffDelay: 0,
+    };
+    this.pool[index] = poolItem;
+    this.apiToIndexMap.set(api, index);
   }
 
-  addBatchToConnections(apis: T[]): void {
-    apis.forEach((api) => this.addToConnections(api));
+  addBatchToConnections(endpointToApiIndex: Record<string, T>): void {
+    for (const endpoint in endpointToApiIndex) {
+      this.addToConnections(endpointToApiIndex[endpoint], endpoint);
+    }
   }
 
   async connectToApi(apiIndex: number): Promise<void> {
     await this.allApi[apiIndex].apiConnect();
+    this.pool[apiIndex].connection = this.allApi[apiIndex];
   }
 
   get api(): T {
     const index = this.getNextConnectedApiIndex();
     if (index === -1) {
-      throw new Error('No connected api');
+      throw new Error('All of the endpoints are suspended at the moment');
```

Comment:
Is this desired behaviour? If all the endpoints are suspended it doesn't mean they're unusable. If we throw here the node will most likely restart and resume with reset weights then get back to the same point and restart again.

Ideally we can distinguish between an endpoint not working and rate limiting. With rate limiting we should just throttle and with endpoints being down we should exit if they are all failing